server/hotrod/src/main/java/org/infinispan/server/hotrod/Authentication.java

```diff
@@ -86,74 +75,109 @@ public void auth(HotRodHeader header, String mech, byte[] response) throws Privi
+               writeException(header, e);
+               return;
             }
          }
-         byte[] serverChallenge = saslServer.evaluateResponse(response);
+      }
+      byte[] serverChallenge;
+      try {
+         serverChallenge = saslServer.evaluateResponse(response);
+      } catch (SaslException e) {
+         dispose(saslServer);
+         writeException(header, e);
+         return;
+      }
 
-         writeResponse(header, header.encoder().authResponse(header, server, channel.alloc(), serverChallenge));
+      if (saslServer.isComplete()) {
+         // Obtaining the subject might be expensive, so do it before sending the final server response, otherwise the
+         // client might send another operation before this is complete
+         List<Principal> extraPrincipals = new ArrayList<>();
+         String id = normalizeAuthorizationId(saslServer.getAuthorizationID());
+         extraPrincipals.add(new SimpleUserPrincipal(id));
+         extraPrincipals.add(new InetAddressPrincipal(((InetSocketAddress) channel.remoteAddress()).getAddress()));
+         SslHandler sslHandler = (SslHandler) channel.pipeline().get("ssl");
+         try {
+            if (sslHandler != null)
+               extraPrincipals.add(sslHandler.engine().getSession().getPeerPrincipal());
+         } catch (SSLPeerUnverifiedException e) {
+            // Ignore any SSLPeerUnverifiedExceptions
+         }
+         subject = callbackHandler.getSubjectUserInfo(extraPrincipals).getSubject();
+      }
 
-         if (saslServer.isComplete()) {
-            List<Principal> extraPrincipals = new ArrayList<>();
-            String id = normalizeAuthorizationId(saslServer.getAuthorizationID());
-            extraPrincipals.add(new SimpleUserPrincipal(id));
-            extraPrincipals.add(new InetAddressPrincipal(((InetSocketAddress) channel.remoteAddress()).getAddress()));
-            SslHandler sslHandler = (SslHandler) channel.pipeline().get("ssl");
-            try {
-               if (sslHandler != null)
-                  extraPrincipals.add(sslHandler.engine().getSession().getPeerPrincipal());
-            } catch (SSLPeerUnverifiedException e) {
-               // Ignore any SSLPeerUnverifiedExceptions
-            }
-            subject = callbackHandler.getSubjectUserInfo(extraPrincipals).getSubject();
-            String qop = (String) saslServer.getNegotiatedProperty(Sasl.QOP);
-            if (qop != null && (qop.equalsIgnoreCase("auth-int") || qop.equalsIgnoreCase("auth-conf"))) {
-               SaslQopHandler qopHandler = new SaslQopHandler(saslServer);
-               channel.pipeline().addBefore("decoder", "saslQop", qopHandler);
-            } else {
-               saslServer.dispose();
-               callbackHandler = null;
-               saslServer = null;
-            }
+      writeResponse(header, header.encoder().authResponse(header, server, channel.alloc(), serverChallenge));
+
+      if (saslServer.isComplete()) {
```

Comment:
Isn't here a race between responding with the authResponse and adding the QOP handler below?

Reply:
@rvansa @tristantarrant isn't this still valid?